### PR TITLE
SP Ironsights #20, Tau Prediction Fix, SP Crossbow Bolts #21

### LINKED ISSUE #20: Singleplayer Crossbow/357 Ironsights
convar exists to smooth zoom, investigate forcing clientside weapon animation for ironsights with SendWeaponAnim vtable func

### PR DESCRIPTION
**Pull request by ampreeT**
Changes:

- Added feature `SP_WEAPONS` that controls whether to use single-player functionality for weapons
- Added ironsights for `weapon_357` and `weapon_crossbow` (Controlled by feature `SP_WEAPONS`)
- Added regular crossbow bolts and changed crossbow skin (Controlled by feature `SP_WEAPONS`)
- Added tau cannon prediction fix (Issue was caused by hooking `CWeapon_Tau::ItemPostFrame()` and setting IsMultiplayer to `false`)

Fixes #20 and #21

image

### CODE REVIEW

**Comment by ampreeT** [comment]
Currently, the only issues are the following:

- Secondary attack on the crossbow does not fire from the center of the screen
- Secondary attack on the crossbow still plays mp sound
- Crossbow suffers from prediction issues caused by `CWeapon_Crossbow::ItemPostFrame()` being hooked, but this is needed currently for SP implementation

I'll make a new issue for the above once this gets merged in.

File: scripting/include/srccoop/blackmesa/entitypatch.inc [hunk 1]
```diff
@@ -791,4 +793,341 @@ public Action Timer_SetTripmineSpriteColor(Handle timer, CSpriteTeam pSpriteTeam
...
 	return Plugin_Continue;
+}
+
+//------------------------------------------------------
+// weapon_357/weapon_crossbow
+// New constructed logic of toggling ironsights.
+//------------------------------------------------------
+public MRESReturn Hook_ToggleIronsights(int _this, Handle hParams)
+{
+	if (CoopManager.IsCoopModeEnabled() && CoopManager.IsFeatureEnabled(FT_SP_WEAPONS))
```
Comment (Alienmario) [inline scripting/include/srccoop/blackmesa/entitypatch.inc:804]: IsCoopModeEnabled is redundant if you check for feature

File: scripting/include/srccoop/blackmesa/entitypatch.inc [hunk 2]
```diff
@@ -791,4 +793,341 @@ public Action Timer_SetTripmineSpriteColor(Handle timer, CSpriteTeam pSpriteTeam
...
+				pPlayer.ShowCrosshair(!bWillBeUsingIronsights);
+			}
+		}
+		pPlayer.SetFOV(pWeapon, bWillBeUsingIronsights ? iFOV : 0, 0.5);
+		pWeapon.SetIronsights(bWillBeUsingIronsights);
+
+		if (bWillBeUsingIronsights)
+		{
+			// TODO: Spam clicking will block input for a long time.
+			// Fix this
```
[the remark this answers is not in the document]
Comment (Alienmario) [reply scripting/include/srccoop/blackmesa/entitypatch.inc:858]: We also have an util available `PreventWeaponFiring(CBaseCombatWeapon pWeapon, float flDelayTime)`

[the remark this answers is not in the document]
Comment (ampreeT) [reply scripting/include/srccoop/blackmesa/entitypatch.inc]: Changed the implementation to `PreventWeaponFiring`.

File: scripting/include/srccoop/blackmesa/entitypatch.inc (hunk 1 above)
Comment (ampreeT) [reply scripting/include/srccoop/blackmesa/entitypatch.inc]: Removed all instances of `CoopManager.IsCoopModeEnabled()` with the feature.

**Pull request merged**